File: app/src/main/java/com/examle/curexchange/data/repository/currency/CurrencyRepository.java
package com.examle.curexchange.data.repository.currency;

import com.examle.curexchange.data.database.dao.CurrencyDao;
import com.examle.curexchange.data.database.entities.CurrencyEntity;
import com.examle.curexchange.data.model.pojo.CryptoCode;
import com.examle.curexchange.data.model.pojo.Row;
import com.examle.curexchange.data.remote.ApiCryptoCode;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableMaybeObserver;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class CurrencyRepository {

    private List<String> names;
    private final List<Row> rows;
    private List<CurrencyEntity> currencyEntities;
    private CurrencyDao currencyDao;
    private final Flowable<CryptoCode> flowableApiCryptoCode;
    private final Flowable<List<String>> flowableDbQuery;
    private boolean isDbEmpty = true;

    @Inject
    public CurrencyRepository(ApiCryptoCode apiCryptoCode, CurrencyDao currencyDao) {
        this.currencyEntities = new ArrayList<>();
        this.names = new ArrayList<>();
        this.rows = new ArrayList<>();
        this.currencyDao = currencyDao;
        this.flowableApiCryptoCode = apiCryptoCode.getCryptoCodes();
        this.flowableDbQuery = currencyDao.queryCryptoNames();
    }

    public Flowable<List<String>> getNames() {
        return queryData();
    }

    private Flowable<List<String>> queryData() {
        return flowableApiCryptoCode
                .subscribeOn(Schedulers.io())
                .flatMap(code -> {
                   if (isDbEmpty()) {
                        insertToDbFromNetwork(code);
                    }
                    return flowableDbQuery
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .doOnNext(entityNames -> names.addAll(entityNames));
                })
                .observeOn(AndroidSchedulers.mainThread());
    }

    private void insertToDbFromNetwork(CryptoCode code) {
        rows.addAll(code.getRows());
        for (int i = 0; i < rows.size(); i++) {
            currencyEntities.add(new CurrencyEntity(rows.get(i).getCode(), rows.get(i).getName()));
        }
        currencyDao.insertAll(currencyEntities);
    }

    private boolean isDbEmpty() {
        currencyDao.queryOneLine()
                .subscribe(new DisposableMaybeObserver<CurrencyEntity>() {
                    @Override
                    public void onSuccess(CurrencyEntity currencyEntity) {
                        if (currencyEntity != null) {
                            isDbEmpty = false;
                        } else {
                            isDbEmpty = true;
                        }
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
        return isDbEmpty;
    }
}

File: app/src/main/java/com/examle/curexchange/data/repository/currency/CurrencyRepository.kt
package com.examle.curexchange.data.repository.currency

import com.examle.curexchange.data.database.dao.CurrencyDao
import com.examle.curexchange.data.database.entities.CurrencyEntity
import com.examle.curexchange.data.model.pojo.CryptoCode
import com.examle.curexchange.data.model.pojo.Row
import com.examle.curexchange.data.remote.ApiCryptoCode
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class CurrencyRepository @Inject constructor(val apiCryptoCode: ApiCryptoCode,
                                             val currencyDao: CurrencyDao) {

    var rows: MutableList<Row> = ArrayList()
    var currencyEntities: MutableList<CurrencyEntity> = ArrayList()

    fun getNames(): Flowable<MutableList<String>> =
            currencyDao.queryOneLine()
                    .map { ignore -> false }
                    .switchIfEmpty(Maybe.just(true))
                    .toFlowable()
                    .subscribeOn(Schedulers.io())
                    .flatMap { isEmpty ->
                        if (isEmpty) {
                            apiCryptoCode.getCryptoCodes()
                                    .flatMap { code ->
                                        insertToDbFromNetwork(code)
                                        currencyDao.queryCryptoNames()
                                    }
                        } else {
                            currencyDao.queryCryptoNames()
                        }
                    }

    fun insertToDbFromNetwork(code: CryptoCode) {
        rows.addAll(code.rows)
        for (i in rows.indices) {
            currencyEntities.add(CurrencyEntity(rows[i].code, rows[i].name))
        }
        currencyDao.insertAll(currencyEntities)
    }
}